federation/pkg/federation-controller/service/dns/dns.go

```diff
@@ -527,21 +527,27 @@ func (s *ServiceDNSController) ensureDNSRecords(clusterName string, service *v1.
 	}
 	commonPrefix := serviceName + "." + namespaceName + "." + s.federationName + ".svc"
 	// dnsNames is the path up the DNS search tree, starting at the leaf
-	dnsNames := []string{
-		strings.Join([]string{commonPrefix, zoneNames[0], regionName, s.serviceDNSSuffix}, "."), // zone level - TODO might need other zone names for multi-zone clusters
-		strings.Join([]string{commonPrefix, regionName, s.serviceDNSSuffix}, "."),               // region level, one up from zone level
-		strings.Join([]string{commonPrefix, s.serviceDNSSuffix}, "."),                           // global level, one up from region level
-		"", // nowhere to go up from global level
+	var dnsNames [][]string
+	for _, zoneName := range zoneNames {
+		zoneDNSNames := []string{
+			strings.Join([]string{commonPrefix, zoneName, regionName, s.serviceDNSSuffix}, "."), // zone level - TODO might need other zone names for multi-zone clusters
+			strings.Join([]string{commonPrefix, regionName, s.serviceDNSSuffix}, "."),           // region level, one up from zone level
+			strings.Join([]string{commonPrefix, s.serviceDNSSuffix}, "."),                       // global level, one up from region level
+			"", // nowhere to go up from global level
+		}
+		dnsNames = append(dnsNames, zoneDNSNames)
```

Comment:
@quinton-hoole Thanks for the feedback. Maybe I do not fully understand the logic, but here `dnsNames` is basically:
```
[
  [zone1, region, global],
  [zone2, region, global],
  [zone3, region, global],
]
```

And each of the three arrays is passed to `ensureDNSRrsets`. Did you mean that the regional and global records will be duplicated?